### PR TITLE
feat(helm): Added volume support for init containers

### PR DESCRIPTION
**Pull request by stevehipwell**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

This PR adds support for providing volumes to init containers, specifically for the use case adding a script via a ConfigMap. This is done via the addition of `additionalVolumes` & `additionalVolumeMounts` inputs which allow for the addition of pod volumes and also support mounting extra volumes to the default container (mainly for consistency).

It also fixes a chart typo in DaemonSet templates for the `additionalHostPathVolumes` which wouldn't have worked as 
 (not) documented. I would also suggest that the `additionalHostPathVolumes` be deprecated as the new `additionalVolumes` input is more flexible and idiomatic.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Release note**:
```release-note

```

### CODE REVIEW

File: helm/provisioner/templates/daemonset_linux.yaml [hunk 1]
```diff
@@ -106,8 +109,11 @@ spec:
           hostPath:
             path: {{ $classConfig.hostDir }}
         {{- end }}
-        {{- range $name, $path := .Values.daemonset.additionalHostPathVolumes }}
```
[the remark this answers is not in the document]
Comment (stevehipwell) [reply helm/provisioner/templates/daemonset_linux.yaml:109]: I think this is a bug fix not a functional change, so v1.1.0 makes sense.

File: helm/provisioner/templates/daemonset_windows.yaml [hunk 2]
```diff
@@ -103,6 +103,9 @@ spec:
               mountPath: \\.\pipe\csi-proxy-volume-v1beta2
             - name: csi-proxy-filesystem-v1beta2
               mountPath: \\.\pipe\csi-proxy-filesystem-v1beta2
+            {{- with .Values.daemonset.additionalVolumeMounts }}
```
[the remark this answers is not in the document]
Comment (stevehipwell) [reply helm/provisioner/templates/daemonset_windows.yaml:106]: I mirrored the existing behaviour of treating the volume mounts the same for both Linux & Windows.

I think there is more foundational work to correctly support/not-support Windows such as being able to disable the Windows DaemonSet as most clusters aren't going to need it.

[the remark this answers is not in the document]
Comment (mauriciopoppe) [reply helm/provisioner/templates/daemonset_windows.yaml:106]: Yes the disable is going to happen with https://github.com/kubernetes-sigs/sig-storage-local-static-provisioner/issues/381

For the values while this is mixed in the past we've added Windows specific values e.g nodeSelector

- Linux https://github.com/kubernetes-sigs/sig-storage-local-static-provisioner/blob/be7c6372d53dc7aa248ebb6738d71e553e1d9dbc/helm/provisioner/templates/daemonset_linux.yaml#L37-L39
- Windows https://github.com/kubernetes-sigs/sig-storage-local-static-provisioner/blob/be7c6372d53dc7aa248ebb6738d71e553e1d9dbc/helm/provisioner/templates/daemonset_windows.yaml#L37-L39

Maybe we should use Windows specific suffixes in the keys too, what do you think about removing the Windows config for now?

[the remark this answers is not in the document]
Comment (mauriciopoppe) [reply helm/provisioner/templates/daemonset_windows.yaml:106]: thanks for the suggestion, I believe a few of these configurations can be flattened for sure, I'm reading a few charts in https://github.com/bitnami/charts/ to see how a standard library looks like.

For v2 I think we could flatten common but I believe `daemonSet` should stay, I took https://github.com/bitnami/charts/blob/main/bitnami/kube-state-metrics/values.yaml as a reference

[the remark this answers is not in the document]
Comment (stevehipwell) [reply helm/provisioner/templates/daemonset_windows.yaml:106]: I wouldn't take Bitnami charts as a reference point as they're at best "non-standard" and live in their own ecosystem. A good single component Helm chart would be just running `helm create` and good example of a multi component chart (with a primary component) would be [Cert Manager](https://artifacthub.io/packages/helm/cert-manager/cert-manager). In the case of this project I think it's a single component with OS flavours.

[the remark this answers is not in the document]
Comment (mauriciopoppe) [reply helm/provisioner/templates/daemonset_windows.yaml:106]: sounds good, I'll take inspiration from https://github.com/cert-manager/cert-manager/blob/master/deploy/charts/cert-manager/values.yaml to flatten the values

**Comment by mauriciopoppe** [comment]
we flattened the values, I'm also thinking `additionalVolumes` and `additionalVolumeMounts` could replace `additionalHostPathVolumes`

**Comment by stevehipwell** [comment]
@mauriciopoppe would you like a new PR based on the changed code to replace `additionalHostPathVolumes` with `additionalVolumes` and `additionalVolumeMounts`?

**Comment by mauriciopoppe** [comment]
@stevehipwell yes and it'd be appreciated, that way this change would only focus on init containers only

**Comment by stevehipwell** [comment]
@mauriciopoppe it might be easier for me to close this PR and open 2 new ones, one for volumes and the other for init containers. Are you OK with that?

**Comment by mauriciopoppe** [comment]
yes that works

**Comment by stevehipwell** [comment]
@mauriciopoppe it looks like you've already added the init container in #393 so it's just the volume refactor. Will we be in a position to release the updated chart once this is done?

**Comment by mauriciopoppe** [comment]
Yeah it'd be the volume refactor only and yes I'll help with the v2 release

**Comment by stevehipwell** [comment]
> Yeah it'd be the volume refactor only and yes I'll help with the v2 release

@mauriciopoppe do you want a PR with just the volume changes to be followed with a separate PR to cut the release then?

FYI did you see #400?

**Comment by mauriciopoppe** [comment]
Sure a different PR or this PR works fine, taking a look at #400

**Comment by stevehipwell** [comment]
This has been superseded by 401.

/close

**Pull request closed without merging**